### PR TITLE
Add Python 3.10 and pytest 7 to CI

### PR DESCRIPTION
**Pull request by utgwkk**
- <del>drop Python 3.6 (as it is [EOL](https://peps.python.org/pep-0494/))</del>
- add Python 3.10

### CODE REVIEW

**Comment by henryiii** [comment]
I would wait on the drop until there's a release with `python_requires` specified. Then pip will back solve to the older working version on 3.6, rather than breaking on 3.6. Also there's not much point in dropping 3.6 without dropping 2.7, which is 2+ years past EoL at this point. 3.6 will still be supported by ubuntu 18.04 and manylinux until next year, while 2.7 has been removed from manylinux2010+ for a while now.

**Comment by utgwkk** [comment]
OK, I stopped dropping Python 3.6 from testing target in this PR. Thank you for your suggestion! @henryiii

**Comment by henryiii** [comment]
(PS: to be clear, I don't mind dropping 2.7 or 3.6 soon, just want a `python_requires` release first so we can be clear about what we support and making sure back solving works :) )

**Comment by utgwkk** [comment]
By the way, testing with Python 3.10 failed with a strange error...
https://github.com/utgwkk/pytest-github-actions-annotate-failures/runs/7159541333?check_suite_focus=true

**Comment by henryiii** [comment]
This one?

```
File "/home/runner/work/pytest-github-actions-annotate-failures/pytest-github-actions-annotate-failures/.tox/py310-pytest4-linux/lib/python3.10/site-packages/_pytest/assertion/rewrite.py", line 406, in _rewrite_test
[70](https://github.com/utgwkk/pytest-github-actions-annotate-failures/runs/7159541333?check_suite_focus=true#step:5:71)
      co = compile(tree, fn.strpath, "exec", dont_inherit=True)
[71](https://github.com/utgwkk/pytest-github-actions-annotate-failures/runs/7159541333?check_suite_focus=true#step:5:72)
  TypeError: required field "lineno" missing from alias
```

**Comment by henryiii** [comment]
Ahh, pytest must be pinned. You need a more recent pytest to support Python 3.10. They got bit by some changes in 3.10 (and again in 3.11).

**Comment by henryiii** [comment]
`6.2.5` was the first to officially support 3.10 (though I think 6.2.4 might have mostly worked, before that `assert <stuff>` was not supported on 3.10).

**Comment by utgwkk** [comment]
I see. Tests has passed in Python 3.10 with pytest 6.2.5.
https://github.com/utgwkk/pytest-github-actions-annotate-failures/runs/7159633865?check_suite_focus=true

**Comment by henryiii** [comment]
Ahh, interesting idea with the all tests pass. Haven't seen that before. Nice way to support chaining tests & required checks.

**Comment by henryiii** [comment]
Pushed a potential fix. I've combined all the pytest jobs (saves about 50% of the time by avoiding rerunning the job setup each time), and skips the invalid combinations.

**Comment by henryiii** [comment]
Also added PyTest 7.

**Review by henryiii — APPROVED** [review]

**Comment by henryiii** [comment]
Nice to see no code needed changing. :)

File: .github/workflows/test.yml
Comment (utgwkk) [inline .github/workflows/test.yml:49]: Pytest >= 5 doesn't support Python 2.7, so like this? (Though `tox` command simply exits without testing with Python 2.7 and pytest 5)
```suggestion
      if: ${{ matrix.python-version != '2.7' && matrix.python-version != '3.10' }}
```

**Comment by utgwkk** [comment]
Thank you! Could you clean commit log before merging? There are some fixup commits left. @henryiii

File: .github/workflows/test.yml
Comment (henryiii) [reply .github/workflows/test.yml:49]: Wow, that's weird on the 2.7 pytest 5 step:

```
Run tox
  tox
  shell: /usr/bin/bash -e {0}
  env:
    pythonLocation: /opt/hostedtoolcache/Python/2.7.18/x64
    PKG_CONFIG_PATH: /opt/hostedtoolcache/Python/2.7.18/x64/lib/pkgconfig
    LD_LIBRARY_PATH: /opt/hostedtoolcache/Python/2.7.18/x64/lib
    PLATFORM: ubuntu-lates
    PYTEST_MAJOR_VERSION: 5
    PYTEST_PLUGINS: pytest_github_actions_annotate_failures
___________________________________ summary ____________________________________
  congratulations :)
```

It does nothing and passes.

**Comment by henryiii** [comment]
Sure, if you don't like squash and merge.

File: tox.ini
Comment (utgwkk) [inline tox.ini:2]: We have to add `pytest7` env for envlist to test with pytest 7.
```suggestion
envlist = py27-pytest4-{linux,windows}, py{36,37,38,39}-pytest{4,5,6,7}-{linux,windows}, py{310}-pytest{6,7}-{linux,windows}
```

File: tox.ini
Comment (utgwkk) [inline tox.ini:22]:
```suggestion
    7: pytest7
```

**Comment by utgwkk** [comment]
Thank you very much!

File: tox.ini
Comment (henryiii) [reply tox.ini]: This is just to run if nothing was specified, right? You can request an environment that's not in the env list manually, like we are doing in CI. (still needs to be added, but it would have been running without this in CI, except for the mistake below, which was causing it to run pytest6 again)

**Pull request merged**
